### ISSUE TITLE
AssertionError in pop_bubbles_coverage_based.py

### ISSUE THREAD

**Issue by ptrebert** [marbl/verkko#39]
Hi,
running `#1a00b60` (after getting #37) of Verkko leads to the following AssertionError for some runs:

```
Traceback (most recent call last):
  File "/repos/verkko/lib/verkko/scripts/pop_bubbles_coverage_based.py", line 267, in <module>
    try_pop_tip(">" + node, edges, coverage, removed_nodes, removed_edges, avg_coverage*.75)
  File "/repos/verkko/lib/verkko/scripts/pop_bubbles_coverage_based.py", line 142, in try_pop_tip
    assert node[1:] in coverage
AssertionError
```
More info:
- the E. coli test dataset is processed successfully with version `#1a00b60` on my system
  - [remark: minor inconsistency in the readme? Presumably, all the `assembly*` files are the expected final output, right!?]
-  the AssertionError only occurs if Verkko is running the read correction, i.e. the same sample can be processed successfully when I supply already error-corrected HiFi reads and run Verkko with `--no-correction`. The corrected and uncorrected HiFi data may not contain identical sets of reads, though.

Best,
Peter

**Comment by skoren** [comment]
Can you tar/upload the contents of the 5-untip folder to reproduce locally? You can send it using the same info in the Canu FAQ: https://canu.readthedocs.io/en/latest/faq.html#how-can-i-send-data-to-you. This will include your genome sequence in homopolymers compressed space so if you don't want to share that, edit the GFAs to change the S lines to have * instead of sequence. 

As far as the output, yes the `assembly*` is the final result but that was added after the beta so we left the README to be accurate for both beta and later results as the assembly files are copies of those the README lists.

**Comment by ptrebert** [comment]
Thanks - the data is up.

**Comment by skoren** [comment]
Thanks, I got the data. It seems there are some files that should be there which aren't present, specifically `5-untip/nodecov_hifi_fix.csv`. Can you try running:
```
echo Fix coverage
 /repos/verkko/lib/verkko/scripts/get_original_coverage.py \
  ../5-untip/unitig-normal-connected-tip.gfa \
  combined-nodemap-1.txt \
  combined-edges-1.gfa \
  nodelens-1.txt \
  ../1-buildGraph/hifi_nodecov.csv \
> ../5-untip/nodecov_hifi_fix.csv
```
and see whether that file is created and is not empty?

Did you update verkko during the run? That tends to cause issues as it tries to continue on even if the previous steps got changed and have different output (e.g.  see #37).

**Comment by ptrebert** [comment]
That file is being generated:

```
node    coverage
utig3-0 4.5
utig3-1 4.5
utig3-2 8.375
utig3-3 10.5
utig3-4 12.5
utig3-5 12.147698999262586
utig3-7 4.0033091088834905
utig3-8 11.8696
utig3-9 10.0
[...]
```

I have containerized the source build of Verkko to make deployment on our cluster simpler, so it's not a wild mix in the code base, but I did not manually delete the previous output assuming the Snakemake workflow would do the right thing - I'll delete everything and check again...

**Comment by ptrebert** [comment]
A few checkmarks first:
- I have now prepended a `rm -rf <workdir>` to the Verkko call to avoid any interference with previous results
- `git rev-parse --short HEAD` gives `#1a00b60` as expected
- the missing file `5-untip/nodecov_hifi_fix.csv` is a consequence of Snakemake's behavior; excerpt from the log:

```
Removing output files of failed job untip since they might be corrupted:
5-untip/nodecov_hifi_fix.csv
```
I extracted the script that actually throws the AssertionError (`pop_bubbles_coverage_based.py`), and executed it manually after adding more info to the assertion. No surprise, though... here is the head* of the file `5-untip/nodecov_hifi_fix.csv`:

```
node    coverage
utig3-0 9.22222
utig3-2 7.72727
utig3-3 10.7755
utig3-4 10.279769169369901
[...]
```
(*different values than above b/c I tested several samples in parallel)

and this is the problem:
```
    assert node[1:] in coverage, f'{node} - {node[1:]}'
AssertionError: >utig3-1 - utig3-1
```
I checked what happens to `utig3-1` in `get_original_coverage.py`, and it is dropped in the final for loop of the script b/c of the condition
```
if node not in original_coverages: continue
```
leading to this
```
if length_sum == 0: continue
```
evaluating to True. The only node triple for `utig3-1` is `('>gapont-7-len-5688-cov-4', 0, 5688)`

Does that help in diagnosing the problem?
